tracepoint_decode/src/perf_event_data.rs

```diff
@@ -0,0 +1,583 @@
+/// information, including the event's fields.
+#[derive(Clone, Debug)]
+pub struct PerfSampleEventInfo<'a> {
+    /// The bytes of the event, including header and content, in event byte order.
+    ///
+    /// The data consists of the 8-byte header followed by the content, both in event byte order.
+    /// The format of the content depends on `header_type`.
+    ///
+    /// Valid always.
+    pub data: &'a [u8],
+
+    /// Information about the session that collected the event, e.g. clock id and
+    /// clock offset.
+    ///
+    /// Valid always.
+    pub session_info: &'a PerfSessionInfo,
+
+    /// Information about the event (shared by all events with the same `id`).
+    ///
+    /// Valid always.
+    pub event_desc: &'a PerfEventDesc,
+
+    /// Valid if `sample_type()` contains `Identifier` or `Id`.
+    pub id: u64,
+
+    /// Valid if `sample_type()` contains `IP`.
+    pub ip: u64,
+
+    /// Valid if `sample_type()` contains `Tid`.
+    pub pid: u32,
+
+    /// Valid if `sample_type()` contains `Tid`.
+    pub tid: u32,
+
+    /// Use SessionInfo.TimeToTimeSpec() to convert to a TimeSpec.
+    ///
+    /// Valid if `sample_type()` contains `Time`.
+    pub time: u64,
+
+    /// Valid if `sample_type()` contains `Addr`.
+    pub addr: u64,
+
+    /// Valid if `sample_type()` contains `StreamId`.
+    pub stream_id: u64,
+
+    /// Valid if `sample_type()` contains `Cpu`.
+    pub cpu: u32,
+
+    /// Valid if `sample_type()` contains `Cpu`.
+    pub cpu_reserved: u32,
+
+    /// Valid if `sample_type()` contains `Period`.
+    pub period: u64,
+
+    /// Read format data.
+    ///
+    /// Valid if `sample_type()` contains `Read`.
+    pub read_range: ops::Range<u16>,
+
+    /// Callchain data.
+    ///
+    /// Valid if `sample_type()` contains `Callchain`.
+    pub callchain_range: ops::Range<u16>,
+
+    /// Raw event data.
+    ///
+    /// Valid if `sample_type()` contains `Raw`.
+    pub raw_range: ops::Range<u16>,
+}
+
+impl<'a> PerfSampleEventInfo<'a> {
+    /// Constructs a new PerfSampleEventInfo instance.
+    /// Requires that `data` is at least 8 bytes long (must start with the [`PerfEventHeader`]).
+    pub const fn new(
+        data: &'a [u8],
+        session_info: &'a PerfSessionInfo,
+        event_desc: &'a PerfEventDesc,
+    ) -> Self {
+        debug_assert!(data.len() >= mem::size_of::<PerfEventHeader>());
+        return Self {
+            data,
+            session_info,
+            event_desc,
+            id: 0,
+            ip: 0,
+            pid: 0,
+            tid: 0,
+            time: 0,
+            addr: 0,
+            stream_id: 0,
+            cpu: 0,
+            cpu_reserved: 0,
+            period: 0,
+            read_range: 0..0,
+            callchain_range: 0..0,
+            raw_range: 0..0,
+        };
+    }
+
+    /// Returns true if the the session's event data is formatted in big-endian
+    /// byte order. (Use `byte_reader()` to do byte-swapping as appropriate.)
```

Comment:
As appropriate? If it's not going to return data in the native format of the consumer, then this comment should probably be a bit more specific.

Reply:
This provides direct access to the event payload as loaded from the perf.data file. For "as appropriate" to be more specific, you have to know how those bytes are to be decoded.

The `byte_reader` provides a helper class that knows whether a data source was "big-endian" or "little-endian". It takes bytes as input and returns integers as output.

If somebody is directly accessing the event payload, it is up to them to figure out when `byte_reader` is appropriate.